### PR TITLE
Fix race condition issues on slower FS

### PR DESCRIPTION
**Pull request by mlucool**
This adds locking before executing commands. While this may add locking in some cases that are not a race (i.e. things that do not need to acquire index.lock), this solves issues related to using this with a git directory on NFS.

### CODE REVIEW

**Comment by mlucool** [comment]
Let me know if you generally like this. Not sure why it can't find `tornado.locks` as it has been there a while: https://www.tornadoweb.org/en/stable/locks.html

File: jupyterlab_git/git.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 import pexpect
 import tornado
```
Comment (telamonian) [inline jupyterlab_git/git.py]:
```suggestion
import tornado
import tornado.locks
```

It looks like the problem in the test is being caused by the fact that `locks` is a seperate module that doesn't get imported by default on `import tornado`. If you're familiar with scipy, it's just like how you would need to do `import scipy.stats` before you can actually reference `scipy.stats` (or whichever scipy module)

**Comment by telamonian** [comment]
Nice work. I like the thinking behind this PR. We will need a better description of the error/issue this is trying to solve (might be nicest if it was written down in a separate issue). I'd also like to better understand what locking (if any) the git CLI itself does.

Ideally, this PR should also include a test to prove that it fixes the problem it's trying to solve. One idea for the test: explicitly lock a resource for ~5 seconds, then immediately start another operation that would create a race condition if locking is not present.

**Comment by mlucool** [comment]
> I'd also like to better understand what locking (if any) the git CLI itself does.

Git locking: https://docs.microsoft.com/en-us/azure/devops/repos/git/git-index-lock?view=azure-devops

Waiting on locks: https://stackoverflow.com/a/36364687

So we can improve this by telling execute which commands care about a lock file or not (e.g. I don't believe status does), but I think thats an over optimization at this point and error prone.

> We will need a better description of the error/issue this is trying to solve
I think https://github.com/jupyterlab/jupyterlab-git/issues/555 is the same issue. Symptoms are at least

> this PR should also include a test to prove that it fixes the problem it's trying to solve

I'll add one

**Comment by mlucool** [comment]
@telamonian do you have an example where execute is tested? `pytest --cov=jupyterlab_git --cov-report html jupyterlab_git` makes it seem like we always mock that out

**Comment by telamonian** [comment]
> do you have an example where execute is tested? `pytest --cov=jupyterlab_git --cov-report html jupyterlab_git` makes it seem like we always mock that out

Hmm, not off the top of my head. Frederic's the expert, since he wrote execute and most of the related code (and the related tests). @fcollonval Any ideas?

**Comment by telamonian** [comment]
I thought of a potential concern: should read-only ops (in particular `git status`) ignore locking? Basically, I'm wondering if there are any potential interactions between this PR and how our UI updates.

**Comment by mlucool** [comment]
Per my comment:
>While this may add locking in some cases that are not a race (i.e. things that do not need to acquire index.lock)

Yes, `git status` will now unnecessarily wait for the lock. We can remove it in some cases trivially (i.e. opt out) but that feels like an pre-mature optimization based on my local testing. If you feel strongly, I'm happy to make this change

**Comment by mlucool** [comment]
@telamonian let me know what you'd like me to do here to get this merged so that it's included in 0.10.

**Comment by fcollonval** [comment]
> 
> 
> > do you have an example where execute is tested? `pytest --cov=jupyterlab_git --cov-report html jupyterlab_git` makes it seem like we always mock that out
> 
> Hmm, not off the top of my head. Frederic's the expert, since he wrote execute and most of the related code (and the related tests). @fcollonval Any ideas?

Hey sorry for the delay. 
Unfortunately as stated in the out-of-date [roadmap](https://github.com/jupyterlab/jupyterlab-git/blob/master/ROADMAP.md). There is no integration test. Meaning that all git command are mocked. But when I introduced the asynchronous calls, I went to mock the `execute` function directly.

But you could test your nice feature by calling directly the `execute` function and mock the `run_in_executor`; the definition to patch is [tornado.ioloop.IOLoop.run_in_executor](https://www.tornadoweb.org/en/stable/ioloop.html#tornado.ioloop.IOLoop.run_in_executor).

**Comment by fcollonval** [comment]
Testing is a bit more complicated that what I say. But that commit shows how to start achieving it: https://github.com/fcollonval/jupyterlab-git/commit/6ede2a8bed6ab9302ba7c62b3dc6018d40ecebbd

File: jupyterlab_git/git.py [hunk 2]
```diff
@@ -85,19 +94,32 @@ def call_subprocess(
...
-            password,
-            cwd,
-            env,
-        )
-    else:
-        current_loop = tornado.ioloop.IOLoop.current()
-        code, output, error = await current_loop.run_in_executor(
-            None, call_subprocess, cmdline, cwd, env
-        )
+    await execution_lock.acquire(timeout=datetime.timedelta(seconds=MAX_WAIT_FOR_EXECUTE_S))
```
Comment (fcollonval) [inline jupyterlab_git/git.py]:
```suggestion
    try:
        await execution_lock.acquire(timeout=datetime.timedelta(seconds=MAX_WAIT_FOR_EXECUTE_S))
    except  tornado.util.TimeoutError:
        return (1, "", "Unable to get the lock on the directory")
```

We should return a proper error message if the lock can not be acquired.

File: jupyterlab_git/git.py [hunk 3]
```diff
@@ -16,7 +18,14 @@
 # See https://git-scm.com/docs/git-config#_syntax for git var syntax
 CONFIG_PATTERN = re.compile(r"(?:^|\n)([\w\-\.]+)\=")
 DEFAULT_REMOTE_NAME = "origin"
+# Ensure on NFS or similar, that we give the .git/index.lock time to be removed
+MAX_WAIT_FOR_LOCK_S = 5
+# How often should we check for the lock above to be free?
+CHECK_LOCK_INTERVAL_S = 0.1
+# How long to wait to be executed or finished your execution before timing out
+MAX_WAIT_FOR_EXECUTE_S = 20
```
Comment (fcollonval) [inline jupyterlab_git/git.py]: There is an inconsistency here. The waiting time for the lock should be greater than the one for execution. Otherwise, for a long run other calls will most likely be dropped due to a failure in acquiring the lock.

**Comment by fcollonval** [comment]
One additional comment:

The argument `cwd` of `execute` should be mandatory, now:
```diff
-    cwd: "Optional[str]" = None,
+    cwd: str,
```

File: jupyterlab_git/git.py [hunk 4]
```diff
@@ -16,7 +18,14 @@
 # See https://git-scm.com/docs/git-config#_syntax for git var syntax
 CONFIG_PATTERN = re.compile(r"(?:^|\n)([\w\-\.]+)\=")
 DEFAULT_REMOTE_NAME = "origin"
+# Ensure on NFS or similar, that we give the .git/index.lock time to be removed
+MAX_WAIT_FOR_LOCK_S = 5
```
Comment (fcollonval) [inline jupyterlab_git/git.py]: Upon better review this one is not used.

**Comment by telamonian** [comment]
I'm targetting this for the next release (which should be 0.20.0). Once this PR is pulled in we'll also backport it to 0.11.0

File: jupyterlab_git/git.py (hunk 4 above)
Comment (mlucool) [reply jupyterlab_git/git.py]: This should have been used. Fixed. The idea is you have 20s to wait for your turn to execute, and when you run, it will wait up to 5s for the `index.lock`.

**Comment by mlucool** [comment]
@fcollonval this should be all set now. I'll admit the unit test is mediocre because we don't have integration testing. It does prove that if a lock exists, it will sleep and simulates that if a lock exists, a command fails.

**Comment by mlucool** [comment]
> I'm targetting this for the next release (which should be 0.20.0). Once this PR is pulled in we'll also backport it to (jlab 1.0 compatible) 0.11.0

There is no need to backport this to 0.11 for me at this time. I'm hoping to be on 2.x in the near future, but I'll let you know if something changes. If you want to backport anyway, it may be a good idea.

File: jupyterlab_git/tests/test_execute.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+                # Simulate answer1 not having released lock
+                lock_exists = True
+                answer2 = await execute(cmd, **kwargs)
+
+                assert executor.call_count == 2
+                assert answer1[0] == 0
+                assert answer2[0] == 1
+                # Did we really try to wait on the lock?
+                sleep.assert_called()
```
Comment (fcollonval) [inline jupyterlab_git/tests/test_execute.py]:
```suggestion
import pytest
from unittest.mock import patch

from jupyterlab_git.git import execute, execution_lock


@pytest.mark.asyncio
async def test_execute_waits_on_index_lock(tmp_path):
    lock_file = tmp_path / ".git/index.lock"
    lock_file.parent.mkdir(parents=True, exist_ok=True)
    lock_file.write_text("")

    async def remove_lock_file(*args):
        assert "unlocked" not in repr(execution_lock)  # Check that the lock is working
        lock_file.unlink()  # Raise an error for missing file
        
    with patch("tornado.gen.sleep") as sleep:
        sleep.side_effect = remove_lock_file   # Remove the lock file instead of sleeping
        
        assert "unlock" in repr(execution_lock)
        cmd = ["git", "status"]
        kwargs = {"cwd": f"{tmp_path!s}"}
        await execute(cmd, **kwargs)
        assert "unlock" in repr(execution_lock)
        
        assert not lock_file.exists()
        sleep.assert_called_once()
```

Comment (mlucool) [reply jupyterlab_git/tests/test_execute.py]: I was avoiding using any disk, but if you think thats ok I'm happy to go with your suggestion. 

FWIW, per above, I don't think git status modifies the lock and we may want to let get status skip having to wait in the future if it does cause some slowness

Comment (fcollonval) [reply jupyterlab_git/tests/test_execute.py]: > I was avoiding using any disk, but if you think thats ok I'm happy to go with your suggestion.

Ja, it is fine. At some point, we should start integration tests those will be much slower.
 
> FWIW, per above, I don't think git status modifies the lock and we may want to let get status skip having to wait in the future if it does cause some slowness

You are correct. I updated the suggestion to restore the `git dummy` call. The goal is not here to check if the command executed correctly or not. So this is fine.

**Comment by kgryte** [comment]
It is not clear to me that we should continually retry until command success in the event that an `index.lock` is already present, as opposed to parsing error messages and displaying a dialog in the UI which instructs a user to retry the command after waiting a specified period of time.

The main issue I see with this PR is the possibility of a race condition. We don't "lock" the UI to prevent further commands, and, from my [reading](https://github.com/jupyterlab/jupyterlab-git/pull/564/files#diff-041cb4cfb7143f3bbcdd07d28dffa92cR107), we run the risk that the pending command could be executed **after** a subsequent command (e.g., another commit).

If we want to add retry logic, we should presumably lock the UI to prevent further interactions. IMO, that is not particularly desirable, in comparison to pushing the retry to the user, which is exactly what happens when using Git at the command-line.

**Comment by mlucool** [comment]
> It is not clear to me that we should continually retry until command success in the event that an index.lock is already present

We only loop on waiting for the lock, not rerunning the command

> we run the risk that the pending command could be executed after a subsequent command (e.g., another commit).

This enforces that each command wait for `execution_lock.acquire`. I assumed, maybe wrongly, they would acquire the lock in FIFO and thus preserve ordering. If this is not true, we should enforce this somehow.

The reason for this commit has less to do with needing retry, but because some of the action from the UI require multiple commands to be sent (e.g. a `git add` then a `git commit`)

**Comment by kgryte** [comment]
@mlucool Yes, you are correct; from my reading, `tornado` uses something akin to a FIFO queue. In which case, the race condition I alluded to earlier is not applicable.

However, another concern is what happens if a user shuts down a kernel before an `index.lock` file is removed and the queued command/commands has/have executed?

Meaning, is it possible that, e.g., one or more pending commands (e.g., one or more commit commands for "saving" changes) could fail to be executed in the event that the kernel closes during the interim waiting for an `index.lock` file to be removed?

If, instead of locking on the backend, where the locking behavior is hidden from the user, the extension erred immediately and displayed an error modal to the user, recommending manually retrying an action, then, IMO, while arguably more laborious, expectations could be better managed.

In short, I am largely concerned about all the ways that this change could cause the extension to fail. Are there edge cases that we are missing?

Lastly, I am leery about hardcoding intervals for wait times. Wait times would seem to be highly particular to a user's environment and the relevant repository.

For example, one of my side projects is rather large, and I frequently encounter the presence of an `index.lock` which sometimes takes upward of 30-60 seconds to disappear, depending on the current resource constraints on my laptop.

Personally, I don't have a good sense as to what could be considered reasonable.

**Comment by mlucool** [comment]
> However, another concern is what happens if a user shuts down a kernel before an index.lock file is removed and the queued command/commands has/have executed?

I thought extension added routes and interactions are not done via a kernel?

All actions that take more than one command which affects `index.lock` (e.g. git add + commit in simple staging mode) fail because they happen too fast on NFS in my case. These would have to be made atomic from the UI POV and the backend would have to enforce waiting and ordering between those. I'll also note that having more than one UI instance means you can have race conditions between them and is another reason to put it into the backend.

> Lastly, I am leery about hardcoding intervals for wait times. Wait times would seem to be highly particular to a user's environment and the relevant repository.

You are right. We should let users change these. There is no "right" amount of time to wait with very different environments. I think we pick a good default in terms of human timeframes and then let advanced users change this. Should we both `MAX_WAIT_FOR_EXECUTE_S` and `MAX_WAIT_FOR_LOCK_S `?

**Comment by saulshanabrook** [comment]
> These would have to be made atomic from the UI POV and the backend would have to enforce waiting and ordering between those. 

It seems to make sense that if we are allowing a long running process to take place on the server, then we should make sure to make this clear to the user. So we should add some way for them to see what actions are being run on the server, or block the UI while one action is being run.

If that could be done after this PR is merged, instead of in this PR itself, that seems like it could be fine as well.

**Comment by mlucool** [comment]
> So we should add some way for them to see what actions are being run on the server, or block the UI while one action is being run.

This already happens for at least some commands (a popup with a spinner). I agree that we should have feedback for all commands to make it more clear to a user that something is being done.

FWIW, in practice, this wait is still subhuman scale for the cases I tested (fast network, NFS, smallish repo). Any case that would be on human scales (e.g. very large repos that @kgryte pointed out), would either fail (e.g. `git add` and `git commit` would step on each other) or take approximately as long as they would from the CLI.

**Comment by kgryte** [comment]
@telamonian You have any thoughts on this?

Proposal:

1.  Modify PR to allow retry timeout configurability.
2.  Follow-up PR to suspend UI interactions until pending command complete (e.g., a modal with a spinner). I believe this should be straightforward, as we just need to enable a UI element until receiving an HTTP response.

The need for the latter is to prevent the user from closing JupyterLab before pending commands have had a chance to complete (e.g., before an `index.lock` file is removed and `git add && git commit` run, thus potentially leading to the discarding of user changes).

**Comment by mlucool** [comment]
I'm happy with @kgryte proposal if that means @telamonian / @fcollonval will accept this PR. It's been stuck for a while, and it would be good wrap up the work here.

**Comment by fcollonval** [comment]
Thanks @mlucool for the patience. I'll merge this and add an issue with the proposal of enhancement from @kgryte.

**Comment by fcollonval** [comment]
@meeseeksdev backport to 0.11.x
